book/testing.md

```diff
@@ -0,0 +1,138 @@
+
+```
+❯ std assert (1 == 2)
+Error:
+  × Assertion failed.
+   ╭─[entry #13:1:1]
+ 1 │ std assert (1 == 2)
+   ·             ───┬──
+   ·                ╰── It is not true.
+   ╰────
+```
+
+Optionally, a message can be set to show the intention of the assert command, what went wrong or what was expected:
+
+```nushell
+❯ std assert ($a == 19) $"The lockout code is wrong, received: ($a)"
+Error:
+  × The lockout code is wrong, received: 13
+   ╭─[entry #25:1:1]
+ 1 │ std assert ($a == 19) $"The lockout code is wrong, received: ($a)"
+   ·             ────┬───
+   ·                 ╰── It is not true.
+   ╰────
+```
+
+There are many assert commands, which behaves exactly as the base one with the proper operator. The additional value for them is the ability for better error messages.
+
+For example this is not so helpful without additional message:
+
+```
+❯ std assert ($b | str contains $a)
+Error:
+  × Assertion failed.
+   ╭─[entry #35:1:1]
+ 1 │ assert ($b | str contains $a)
+   ·              ──────┬─────
+   ·                    ╰── It is not true.
+   ╰────
+```
+
+While with using `assert str contains`:
+
+```
+❯ std assert str contains $b $a
+Error:
+  × Assertion failed.
+   ╭─[entry #34:1:1]
+ 1 │ assert str contains $b $a
+   ·                     ──┬──
+   ·                       ╰── 'haystack' does not contain 'a needle'.
+   ╰────
+```
+
+In general for base `assert` command it is encouraged to always provide the additional message to show what went wrong. If you cannot use any built-in assert command, you can create a custom one with passing the label for [`error make`](error_make.md) for the `assert` command:
+
+```
+def "assert even" [number: int] {
+    std assert ($number mod 2 == 0) --error-label {
+        start: (metadata $number).span.start,
+        end: (metadata $number).span.end,
+        text: $"($number) is not an even number",
+    }
+}
+```
+
+Then you'll have your detailed custom error message:
+
+```
+❯ let $a = 13
+❯ assert even $a
+Error:
+  × Assertion failed.
+   ╭─[entry #37:1:1]
+ 1 │ assert even $a
+   ·             ─┬
+   ·              ╰── 13 is not an even number
+   ╰────
+```
+
+## Test modules & test cases
+
+The naming convention for test modules is `test_<your_module>.nu`. This module will be sourced many times during testing, so it is strongly discouraged to put any _main_ code there.
+
+The test case commands must be exported commands with `test_<test name>` naming convention, without any required parameters.
+
+For example, `test_math.nu`:
+
+```nushell
+use std.nu *
+
+export def test_addition [] {
+    assert equal (1 + 2) 3
+}
+
+export def test_multiplication [] {
+    assert equal (2 * 3) 6
+}
+
+export def test_failing [] {
+    assert false "This is just for testing"
+}
+```
+
+## Test runner
+
+You can find a test runner in the [standard library folder in Git repository](https://github.com/nushell/nushell/tree/main/crates/nu-utils/standard_library).
```

Comment:
maybe directly in the
```
[`tests` module of the standard library](https://github.com/nushell/nushell/tree/main/crates/nu-utils/standard_library/tests.nu)
```
?

Reply:
I hope I will be able to move the test runner into `std.nu` to be distributed with Nushell. 😋

Also `tests` module of the standard library does not exists, since standard library is what will be distributed with Nushell. At least for me.

I consider this chapter as a preview, which will be updated later when these things get stabilized.

Reply:
yeah okey, these are good points :+1: 

feel free to keep the more general path to the directory :ok_hand: